### PR TITLE
optimization of the roundrobin implementation.

### PR DESCRIPTION
**Pull request by ethanvc**
this mr is going to optimize roundrobin balancer implementation.

benchmark result on my computer:
```shell
$ go test -bench .
goos: darwin
goarch: arm64
pkg: google.golang.org/grpc/balancer/roundrobin
BenchmarkPickConn10-10          16505442                78.84 ns/op
BenchmarkPickConn100-10         15411775                79.29 ns/op
PASS
ok      google.golang.org/grpc/balancer/roundrobin      2.826s
```
Original serials is the old implementation, Fast serials is the new one. we can see the new one is more than 38% faster.

RELEASE NOTES: N/A

### CODE REVIEW

**Comment by ethanvc** [comment]
@zasweq Thanks for your suggestion. Here i added test and benchmark related code.

**Comment by ethanvc** [comment]
@zasweq hello, how long will this pr merge to the master? This is my first pr, I do not know the flow of the rest stage.

File: balancer/roundrobin/roundrobin_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package roundrobin
```
Comment (zasweq) [inline balancer/roundrobin/roundrobin_test.go:1]: Thanks for checking this in, but I mainly wanted to see the benchmarking code. I don't really desire to merge this into our main respository.

File: balancer/roundrobin/roundrobin.go [hunk 2]
```diff
@@ -69,15 +69,13 @@ type rrPicker struct {
...
+	next     uint32
 }
 
 func (p *rrPicker) Pick(balancer.PickInfo) (balancer.PickResult, error) {
-	p.mu.Lock()
-	sc := p.subConns[p.next]
-	p.next = (p.next + 1) % len(p.subConns)
-	p.mu.Unlock()
+	subConnsLen := uint32(len(p.subConns))
+	consumed := atomic.AddUint32(&p.next, 1)
```
Comment (zasweq) [inline balancer/roundrobin/roundrobin.go:77]: Nit: I don't like the name consumed here. s/next || nextIndex or something along those lines.

**Review by zasweq — CHANGES_REQUESTED** [review]
Sorry for the late pass.

File: balancer/roundrobin/roundrobin_test.go (hunk 1 above)
Comment (ethanvc) [reply balancer/roundrobin/roundrobin_test.go:1]: The benchmark code is below in this file, the function name is BenchmarkPickConn100. or you want other benchmark test?

File: balancer/roundrobin/roundrobin.go (hunk 2 above)
Comment (ethanvc) [reply balancer/roundrobin/roundrobin.go:77]: yes, i agree with you

File: balancer/roundrobin/roundrobin_test.go (hunk 1 above)
Comment (ethanvc) [reply balancer/roundrobin/roundrobin_test.go:1]: so since you already see the benchmark code, now i just remove it now?

Comment (zasweq) [reply balancer/roundrobin/roundrobin_test.go:1]: Yes please.

File: balancer/roundrobin/roundrobin.go (hunk 2 above)
Comment (zasweq) [reply balancer/roundrobin/roundrobin.go]: Thanks for switching.

**Review by zasweq — CHANGES_REQUESTED** [review]
Yeah just remove that file and I'll approve.

**Comment by ethanvc** [comment]
> 

already removed.

File: balancer/roundrobin/roundrobin_test.go (hunk 1 above)
Comment (zasweq) [reply balancer/roundrobin/roundrobin_test.go:1]: Can you delete this whole file please?

Comment (ethanvc) [reply balancer/roundrobin/roundrobin_test.go]: > Can you delete this whole file please?

done

**Review by zasweq — APPROVED** [review]
LGTM. 

As per my team lead Doug Fawley from other PR:
"This is probably fine as-is. It's not likely to make any difference in a real-world scenario, but isn't any more complicated than the lock-based approach, and it objectively better in performance in a performance-sensitive area.

The unusual ordering once we reach 4B picks is unfortunate, but probably harmless, and would be even more unlikely to occur once we reset transports as they approach the max stream ID limit."

**Comment by littlejiancc** [comment]
After the change, the first connection obtained is `next+1`, and the previous one is `next`.
The starting point of the rotation training is different.
@zasweq

**Comment by zasweq** [comment]
I don't think that's right. Both the original and new change should start at the same place.

**Comment by littlejiancc** [comment]
The behavior of a random index is indeed the same.
I just mean like this:

```
subConnsLen := uint32(len(p.subConns))
sc := p.subConns[p.next%subConnsLen]

atomic.AddUint32(&p.next, 1)
return balancer.PickResult{SubConn: sc}, nil
```

**Pull request merged**